File: cloud/s3/s3_utils.js
/**
 * Install:
 *   npm install aws-sdk
 *
 * Usage:
 *   node s3_utils.js  # for testing
 *
 *   // for programming
 *   const {S3Client} = require('./s3_utils.js');
 *   const client = S3Client('127.0.0.1:8000', 'accessKey1', 'verySecretKey1', {sslEnabled: false});
 *
 * */
const aws = require('aws-sdk');

// NOTE: only one instance is allowed at the same time
const create_s3_client = (url, access_key, secret_key, options) => {
    options = options || {};
    const s3 = new aws.S3({
        accessKeyId: access_key,
        secretAccessKey: secret_key,
        endpoint: url,
        s3ForcePathStyle: true,
        ...options
    });
    return s3;
};

const wrap_s3_func = (f, params) => {
    if (!params) {
        return new Promise((res, rej) => {
            f((err, resp) => {
                if (err) rej(err);
                else res(resp);
            });
        });
    } else {
        return new Promise((res, rej) => {
            f(params, (err, resp) => {
                if (err) rej(err);
                else res(resp);
            });
        });
    }
};

const put_object = (client, bucket, key, data) => {
    const params = {Bucket: bucket, Key: key, Body: data};
    return wrap_s3_func(client.putObject.bind(client), params);
};

const list_buckets = client => {
    return wrap_s3_func(client.listBuckets.bind(client)).then(r => r['Buckets']);
};

const create_bucket = (client, bucket) => {
    return wrap_s3_func(client.createBucket.bind(client), {Bucket: bucket});
};

const list_objects = (client, bucket) => {
    return wrap_s3_func(client.listObjects.bind(client), {Bucket: bucket}).then(r => r['Contents']);
};

const delete_object = (client, bucket, key) => {
    return wrap_s3_func(client.deleteObject.bind(client), {Bucket: bucket, Key: key});
};

const clear_bucket = async (client, bucket) => {
    const objects = await list_objects(client, bucket);
    for (let object of objects) {
        await delete_object(client, bucket, object.Key);
    }
    return objects;
};

const delete_bucket = (client, bucket) => {
    return wrap_s3_func(client.deleteBucket.bind(client), {Bucket: bucket});
};

// NOTE: only one instance is allowed at the same time
class S3Client {
    constructor(url, access_key, secret_key, options) {
        this.client = create_s3_client(url, access_key, secret_key, options);
        const functions = [
            create_bucket,
            list_buckets,
            delete_bucket,
            clear_bucket,
            put_object,
            list_objects,
            delete_object
        ];
        const arguments_to_list = args => {
            const args_ = [];
            for (let arg of args) {
                args_.push(arg);
            }
            return args_;
        };
        const that = this;
        functions.forEach(f => {
            that[f.name] = function () {
                return f(that.client, ...arguments_to_list(arguments));
            }
        });
    }
}

const test_s3_server = async (url, access_key, secret_key, options) => {
    client = new S3Client(url, access_key, secret_key, options);
    const bucket = 'abc';
    const key = 'key' + Math.random()
    try {
        await client.create_bucket(bucket);
    } catch(e) {}
    await client.list_buckets().then('list_buckets: ', r => console.log(r));
    await client.put_object(bucket, key, 'abcdef').then(r => console.log('put_object: ', r));
    await client.list_objects(bucket).then(r => console.log('list_objects: ', r));
    await client.delete_object(bucket, key).then(r => console.log('delete_object: ', r));
    await client.list_objects(bucket).then(r => console.log('list_objects: ', r));
    await client.clear_bucket(bucket).then(r => console.log('clear_bucket: ', r));
    await client.delete_bucket(bucket).then(r => console.log('delete_bucket: ', r)).catch(e => console.log(e));
};

const test_cloudserver = () => test_s3_server('127.0.0.1:8000', 'accessKey1', 'verySecretKey1', {sslEnabled: false});
const test_minio = () => test_s3_server('127.0.0.1:9000', 'minioadmin', 'minioadmin', {sslEnabled: false});

const test = async () => {
    await test_cloudserver();
    await test_minio();
};

module.exports = {
    S3Client
};

if (typeof require !== 'undefined' && require.main === module) {
    test()
}

File: cloud/s3/s3_utils.py
import re
from functools import wraps
import boto3

def create_s3_client(url, access_key, secret_key):
    return boto3.client('s3',
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key,
        endpoint_url = url 
        )


def _get_status_code(resp):
    return resp['ResponseMetadata']['HTTPStatusCode']

def create_bucket(client, bucket_name):
    try:
        return client.create_bucket(Bucket=bucket_name)
    except Exception as e:
        print(e)
        return None

def clear_bucket(client, bucket_name):
    objects = list_objects(client, bucket_name)
    for obj in objects:
        key = obj['Key']
        delete_object(client, bucket_name, key)
    return objects

def delete_bucket(client, bucket_name):
    try:
        return client.delete_bucket(Bucket=bucket_name)
    except Exception as e:
        print(e)
        return None

def list_buckets(client):
    resp = client.list_buckets()
    return resp.get('Buckets', [])

def list_objects(client, bucket_name):
    resp = client.list_objects(Bucket=bucket_name)
    return resp.get('Contents', [])

def put_object(client, bucket_name, key, binary_data):
    return client.put_object(Bucket=bucket_name, Key=key, Body=binary_data)

def get_object(client, bucket_name, key):
    return client.get_object(Bucket=bucket_name, Key=key)

def get_object_body(client, bucket_name, key):
    '''
    s3://bucket_name/key => bytes
    '''
    return client.get_object(Bucket=bucket_name, Key=key)['Body'].read()

def delete_object(client, bucket_name, key):
    '''
    s3://bucket_name/key => bool
    '''
    resp = client.delete_object(Bucket=bucket_name, Key=key)
    return resp and _get_status_code(resp) < 300

def upload_file(client, bucket_name, filename):
    key = re.sub(r'\.', '', filename)
    if key[0] == '/':
        key = key[1:]
    client.upload_file(Bucket=bucket_name, Key=key, Filename=filename)
    return key

class S3Client:
    def __init__(self, url, access_key, secret_key):
        def wrap(f):
            @wraps(f)
            def wrapped(*args, **kwargs):
                return f(self.client, *args, **kwargs)
            return wrapped
        functions = [
                create_bucket,
                clear_bucket,
                delete_bucket,
                list_buckets,
                list_objects,
                put_object,
                get_object,
                get_object_body,
                delete_object,
                upload_file
                ]
        self.__dict__ = {}
        for f in functions:
            self.__dict__[f.__name__] = wrap(f)
        self.url = url
        self.access_key = access_key
        self.secret_key = secret_key
        self.client = create_s3_client(url, access_key, secret_key)

def test_s3_server(url, access_key, secret_key):
    s3 = S3Client(url, access_key, secret_key)
    bucket = 'abc'
    print(s3.create_bucket(bucket))
    print(s3.put_object(bucket, 'aaa.dat', bytes([1,2])))
    print(s3.list_buckets())
    print(s3.list_objects(bucket))
    print(s3.get_object_body(bucket, 'aaa.dat'))
    key = s3.upload_file(bucket, './s3_utils.py')
    print(s3.delete_object(bucket, key))
    print(s3.delete_bucket(bucket))
    print(s3.clear_bucket(bucket))
    print(s3.delete_bucket(bucket))

def test_minio():
    '''
    docker pull minio/minio
    docker run -p 9000:9000 minio/minio server /data
    '''
    test_s3_server('http://localhost:9000', 'minioadmin', 'minioadmin')

def test_cloudserver():
    '''
    git clone https://github.com/scality/S3.git
    cd S3
    yarn install --frozen-lock-file
    yarn start

    or

    docker pull zenko/cloudserver
    docker run -p 8000:8000 -e S3BACKEND=mem zenko/cloudserver
    '''
    test_s3_server('http://localhost:8000', 'accessKey1', 'verySecretKey1')

if __name__ == '__main__':
    test_minio()
    test_cloudserver()

